### PR TITLE
feat: enable write retry and nack pending writes on reconnect

### PR DESCRIPTION
**Pull request by alvarowolfx**
This PR fix an issue that can happen when an error is returned by the service side and we trigger a reconnect, in some cases where a in-flight/pending write is still waiting for response can be left without any ack/nack. 

Towards internal b/329875851

### CODE REVIEW

File: src/managedwriter/stream_connection.ts [hunk 1]
```diff
@@ -107,20 +108,19 @@ export class StreamConnection extends EventEmitter {
   private handleError = (err: gax.GoogleError) => {
     this.trace('on error', err, JSON.stringify(err));
     if (this.shouldReconnect(err)) {
+      err.message = 'reconnect triggered due to: ' + err.message;
+      this.ackAllPendingWrites(err);
       this.reconnect();
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:113]: We no longer need this reconnect?

File: src/managedwriter/stream_connection.ts [hunk 2]
```diff
@@ -107,20 +108,19 @@ export class StreamConnection extends EventEmitter {
   private handleError = (err: gax.GoogleError) => {
     this.trace('on error', err, JSON.stringify(err));
     if (this.shouldReconnect(err)) {
+      err.message = 'reconnect triggered due to: ' + err.message;
+      this.ackAllPendingWrites(err);
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:112]: Instead of NACK, could we resend the request by calling send?

File: src/managedwriter/stream_connection.ts [hunk 3]
```diff
@@ -107,20 +108,24 @@ export class StreamConnection extends EventEmitter {
   private handleError = (err: gax.GoogleError) => {
     this.trace('on error', err, JSON.stringify(err));
     if (this.shouldReconnect(err)) {
+      const retrySettings = this._writeClient['_retrySettings'];
       this.reconnect();
+      if (retrySettings.enableWriteRetries) {
+        this.resendAllPendingWrites();
+      } else {
+        err.message = 'reconnect triggered due to: ' + err.message;
```
Comment (shollyman) [inline src/managedwriter/stream_connection.ts:116]: nit:  maybe something more developer friendly like "connection failed to service before response was received"?   Reconnect is mostly relevant to subsequent writes.

File: src/managedwriter/stream_connection.ts [hunk 4]
```diff
@@ -245,6 +253,27 @@ export class StreamConnection extends EventEmitter {
     return null;
   }
 
+  private resendAllPendingWrites() {
+    let pw = this._pendingWrites.pop();
+    while (pw) {
+      this.send(pw);
```
Comment (shollyman) [inline src/managedwriter/stream_connection.ts:259]: what's the behavior if send fails?  do you need to mark the write bad?

Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:259]: after a number of attempts it is marked as failed here:
* https://github.com/googleapis/nodejs-bigquery-storage/blob/474e2377b22a2da33167097a6e1f6409ec081774/src/managedwriter/stream_connection.ts#L319

and if write retries are not enabled, it is nack here:
* https://github.com/googleapis/nodejs-bigquery-storage/blob/474e2377b22a2da33167097a6e1f6409ec081774/src/managedwriter/stream_connection.ts#L117

File: src/managedwriter/stream_connection.ts (hunk 3 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:116]: yeah, I was already feeling like the wording was not nice here

File: src/managedwriter/writer_client.ts [hunk 5]
```diff
@@ -102,6 +111,15 @@ export class WriterClient {
     return this._open;
   }
 
+  // Enables StreamConnections to automatically retry failed appends.
+  //
+  // Enabling retries is best suited for cases where users want to achieve at-least-once
+  // append semantics. Use of automatic retries may complicate patterns where the user
+  // is designing for exactly-once append semantics.
+  enableWriteRetries(enable: boolean) {
```
Comment (yirutang) [inline src/managedwriter/writer_client.ts:119]: I would remove this and enable it by default, the customer could disable it by setting the maxNumRetry to be 0.
_retrySettings can also be confusing because you don't support all request level retries, you only support resend the request when the connection is broken. Maybe cal the member of maxRetryAttemptsOnFailedConnection? It is a mouthful, not sure if there is better name for it.

File: system-test/managed_writer_client_test.ts [hunk 6]
```diff
@@ -857,6 +862,155 @@ describe('managedwriter.WriterClient', () => {
...
+      await bigquery
+        .dataset(flakyDatasetId)
+        .delete({force: true})
+        .catch(console.warn);
+    });
+
+    it('should manage to send data in scenario where every 10 request drops the connection', async () => {
+      bqWriteClient.initialize();
+      const client = new WriterClient();
+      client.enableWriteRetries(true);
```
Comment (yirutang) [inline system-test/managed_writer_client_test.ts:904]: You should also be able to test the nonretried behavior.

File: src/managedwriter/stream_connection.ts (hunk 1 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:112]: Yeah, it will still be a very bad problem if we cannot distinguish connection level error vs request level error. No matter what we do, it could cause more duplications if we have request level error. We should put a comment here explaining the situation and revisit this (maybe one day the grpc layer will be fixed).

For now, let's remove gax.Status.RESOURCE_EXHAUSTED, from the retry code.

File: src/managedwriter/stream_connection.ts (hunk 3 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:116]: I would say message aborted due to failed connection, please retry the request.

File: system-test/managed_writer_client_test.ts [hunk 7]
```diff
@@ -857,6 +862,155 @@ describe('managedwriter.WriterClient', () => {
...
+    }).timeout(2 * 60 * 1000);
+
+    it('should manage to send data in scenario where opening the connection can fail more frequently', async () => {
+      bqWriteClient.initialize();
+      const client = new WriterClient();
+      client.enableWriteRetries(true);
+      client.setClient(bqWriteClient);
+
+      try {
+        const flakyTableId = generateUuid() + '_initial_connect_failure';
```
[the remark this answers is not in the document]
Comment (alvarowolfx) [reply system-test/managed_writer_client_test.ts:962]: the failure is being triggered, I can see the logs locally and if I disable the retries on those tests, they fail

File: src/managedwriter/stream_connection.ts [hunk 8]
```diff
@@ -87,7 +87,16 @@ export class StreamConnection extends EventEmitter {
     this._connection.on('error', this.handleError);
     this._connection.on('close', () => {
       this.trace('connection closed');
-      this.close();
+      this.reconnect();
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:90]: Let's try not to reconnect for this particular error here: Aborted with an error message of "Closing the stream because it has been inactive for"

File: src/managedwriter/stream_connection.ts [hunk 9]
```diff
@@ -87,6 +87,16 @@ export class StreamConnection extends EventEmitter {
     this._connection.on('error', this.handleError);
     this._connection.on('close', () => {
       this.trace('connection closed');
+      this.reconnect();
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:90]: Reconnect only when the error is retryable?

File: src/managedwriter/stream_connection.ts (hunk 8 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts]: now we only reconnect when there are pending writes, so in the case of idle connection, the reconnection is not gonna happen right away, only on the next write.

File: src/managedwriter/stream_connection.ts (hunk 9 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts]: now we only reconnect when there are pending writes, which will cover the case where the error is retryable.

File: src/managedwriter/stream_connection.ts (hunk 1 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts]: I moved the logic to reconnect only when the connection is closed and there are pending writes and the check if an error is retryable, I removed the `gax.Status.RESOURCE_EXHAUSTED` status.

File: src/managedwriter/writer_client.ts (hunk 5 above)
Comment (alvarowolfx) [reply src/managedwriter/writer_client.ts:119]: I'm afraid of leaving the write retries as ON by default because it is not recommended on a write-once scenario, unless writes are send in sequence and synchronously. Any thoughts @shollyman ?

File: src/managedwriter/stream_connection.ts (hunk 8 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts]: Cool

File: src/managedwriter/stream_connection.ts [hunk 10]
```diff
@@ -106,44 +116,50 @@ export class StreamConnection extends EventEmitter {
...
       );
-      this.ackNextPendingWrite(err);
+      if (!this.isRetryableError(err)) {
+        this.ackNextPendingWrite(err);
+        return;
+      }
+      const retrySettings = this._writeClient['_retrySettings'];
+      if (!retrySettings.enableWriteRetries) {
+        this.ackNextPendingWrite(err);
+      }
```
[the remark this answers is not in the document]
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:141]: if retry is disabled, the pending write is nack'ed with an error and the user has to retry on their own.

[the remark this answers is not in the document]
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:141]: When you got to line 144, retry is already enabled?

File: src/managedwriter/stream_connection.ts [hunk 11]
```diff
@@ -106,44 +118,53 @@ export class StreamConnection extends EventEmitter {
...
+        });
+      } else {
+        this.ackNextPendingWrite(err);
+      }
+      return;
+    }
+    if (this.isRetryableError(err) && this.listenerCount('error') === 0) {
       return;
     }
     this.emit('error', err);
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:147]: This line can be moved to L140? And the block of:
if (this.isPermanentError(err)) {
      this.trace('found permanent error', err);
      while (nextPendingWrite) {
        this.ackNextPendingWrite(err);
        nextPendingWrite = this.getNextPendingWrite();
      }
      this.ackAllPendingWrites(err);
      this.emit('error', err);
      return;
    }

Can be removed.

File: src/managedwriter/stream_connection.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:141]: retry on transient quota error is a common request, for you, it can simply be done by adding
(isRetryableError(error) || err is RE) && enableRetry

File: src/managedwriter/writer_client.ts (hunk 5 above)
Comment (yirutang) [reply src/managedwriter/writer_client.ts:119]: But do you need to allow users to be able to configure retry times?

File: src/managedwriter/stream_connection.ts [hunk 12]
```diff
@@ -280,16 +324,19 @@ export class StreamConnection extends EventEmitter {
...
-      pw._markDone(new Error('connection closed'));
-      return;
-    }
-    if (this._connection.destroyed || this._connection.closed) {
+    if (this.isConnectionClosed()) {
       this.reconnect();
     }
     this.trace('sending pending write', pw);
+    const retrySettings = this._writeClient['_retrySettings'];
+    const tries = pw._increaseRetryAttempts();
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:332]: move this before reconnect?

File: src/managedwriter/stream_connection.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:141]: Thinking more, let's leave ResourceExhausted out this time, if we do a retry it needs to be backoff retry.

[the remark this answers is not in the document]
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:141]: Resolved.

File: src/managedwriter/stream_connection.ts (hunk 11 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:147]: what is the intention here ? nack each pending write individual in case of a permanent request error ? or just remove the concept of permanent errors ?

Comment (yirutang) [reply src/managedwriter/stream_connection.ts:140]: If it is not retryable error then it is a permenant error. :-)

File: src/managedwriter/stream_connection.ts (hunk 12 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:304]: If maxRetryAttempts is 0, does it mean we should send the request? So it should be tries >= retrySettings.maxRetryAttempts?

File: src/managedwriter/stream_connection.ts (hunk 11 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:140]: resolve.

**Review by yirutang — APPROVED** [review]
Please address: https://screenshot.googleplex.com/6RkanhNsPQAGBby and then we are good.

File: src/managedwriter/stream_connection.ts (hunk 12 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:304]: if we change to `>=`, on the first time the request is sent ( with no retries yet ) it is going to throw a error. So I think it should still be just `>`.

Comment (yirutang) [reply src/managedwriter/stream_connection.ts:303]: Resolve.

File: src/managedwriter/stream_connection.ts [hunk 13]
```diff
@@ -87,6 +86,18 @@ export class StreamConnection extends EventEmitter {
     this._connection.on('error', this.handleError);
     this._connection.on('close', () => {
       this.trace('connection closed');
+      if (this.hasPendingWrites()) {
+        this.reconnect();
```
Comment (shollyman) [inline src/managedwriter/stream_connection.ts:90]: does StreamConnection have a notion of terminal error where we wouldn't want reconnect to happen?  I see there's some notion of permanent error further down, but I may not be grokking the flow here.

File: system-test/managed_writer_client_test.ts [hunk 14]
```diff
@@ -857,6 +865,266 @@ describe('managedwriter.WriterClient', () => {
...
+          });
+
+          const pendingWrites: PendingWrite[] = [];
+          const iterations = new Array(50).fill(1);
+          let offset = 10;
+          for (const _ of iterations) {
+            const rows = generateRows(10);
+            const pw = writer.appendRows(rows);
+            pendingWrites.push(pw);
+            offset += 10;
```
Comment (shollyman) [inline system-test/managed_writer_client_test.ts:1044]: lint seems correct here, it's not clear if you intended to use this for an assertion of if you mean to do offset-based appends.

**Review by shollyman — COMMENTED** [review]
there's a few lint warnings about unused loop values that may deserve some scrutiny

File: system-test/managed_writer_client_test.ts (hunk 14 above)
Comment (alvarowolfx) [reply system-test/managed_writer_client_test.ts:1044]: good catch, I had to do some changes to be able to make it work properly on an async scenario.

**Comment by alvarowolfx** [comment]
Summary of integration/e2e tests added on this PR:

## Test Group: Flaky Scenarios

### Test SubGroup: should manage to send data in sequence scenario 
  * Tries to send 50 AppendRows requests with 10 rows each.
  * Requests are sent in sequence - wait for response for current one to send the next.
  * Uses PENDING Stream with offset
  * Validate that 500 rows are inserted by calling finalize on the stream and checking rowCount.

   #### Test Name: every 10 request drops the connection 
    * Uses `reconnect_on_close` table ID, which makes connection to be closed after 10 requests

   #### Test Name: opening the connection can fail more frequently   
    * After every successful Append, we close the connection.  
    * Uses `initial_connect_failure` table ID, which makes opening a connection fail 60% of the time.

### Test SubGroup: should manage to send data in parallel.
  * Tries to send 50 AppendRows requests with 10 rows each.
  * Requests are sent all at once asynchronously and we await for all to finish.
  * Uses PENDING Stream with offset
  * Validate that 500 rows are inserted by calling finalize on the stream and checking rowCount.

  #### Test Name: every 10 request drops the connection:
    * Uses `reconnect_on_close` table ID, which makes connection to be closed after 10 requests.
    
  #### Test Name: every 10 request there is a quota error:
    * Simulates the service side throwing in stream response error.
    * Throws `quota error` + grpc status RESOURCE_EXHAUSTION every 10 requests.

File: src/managedwriter/stream_connection.ts [hunk 15]
```diff
@@ -106,62 +117,50 @@ export class StreamConnection extends EventEmitter {
...
+
+  private isRetryableError(err: gax.GoogleError): boolean {
+    const errorCodes = [
       gax.Status.ABORTED,
+      gax.Status.UNAVAILABLE,
       gax.Status.CANCELLED,
-      gax.Status.DEADLINE_EXCEEDED,
       gax.Status.INTERNAL,
+      gax.Status.DEADLINE_EXCEEDED,
+      gax.Status.RESOURCE_EXHAUSTED,
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:154]: I think we discussed about it and decided to remove this from the retry code for now.

Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:154]: then I'll remove the integration test for in stream RE error

File: src/managedwriter/stream_connection.ts [hunk 16]
```diff
@@ -106,62 +117,49 @@ export class StreamConnection extends EventEmitter {
...
-    const reconnectionErrorCodes = [
-      gax.Status.UNAVAILABLE,
-      gax.Status.RESOURCE_EXHAUSTED,
+  private handleRetry(err: gax.GoogleError) {
+    const retrySettings = this._writeClient['_retrySettings'];
+    if (this.isRetryableError(err) && retrySettings.enableWriteRetries) {
+      if (!this.isConnectionClosed()) {
+        const pw = this._pendingWrites.pop()!;
+        this.send(pw);
+      }
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:141]: In the rare case that conneciton is closed, we didn't send the request again, then we should make sure we nack the request.

I think we should do
if (retry) {
   if (connection not closed) {
       send;
       return;
    }   
}
nack;

File: src/managedwriter/stream_connection.ts (hunk 13 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:90]: Yeah, I think it makes sense to only reconnect when there is pending requests && it is retrable error.

Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:90]: the problem is that the 'close' callback doesn't give the error that caused it. I can create an instance variable on keep track of the last seen error, but not sure how safe this is.

File: src/managedwriter/stream_connection.ts (hunk 16 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:141]: the nack is going be handled on the `'close'` callback in that scenario.

File: src/managedwriter/stream_connection.ts (hunk 13 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts]: ouch...
I think it is still safe here, if we reconnect on permenant error and resend all the request, the connection will be shutdown, but all the pending requests will eventually fail with a limited number of retries. So the pending queue will eventually drain. Also, I think it is worth a logging here, just say we are reconnecting with X number of pending writes.

File: src/managedwriter/stream_connection.ts [hunk 17]
```diff
@@ -183,15 +190,23 @@ export class StreamConnection extends EventEmitter {
   }
 
   private handleData = (response: AppendRowsResponse) => {
-    this.trace('data arrived', response);
-    const pw = this.getNextPendingWrite();
-    if (!pw) {
+    this.trace('data arrived', response, this._pendingWrites.length);
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:193]: This is probably too much logging, it will log one line per request. If it is debug logging then it is fine.

File: src/managedwriter/stream_connection.ts [hunk 18]
```diff
@@ -183,15 +190,23 @@ export class StreamConnection extends EventEmitter {
...
+    if (!this.hasPendingWrites()) {
       this.trace('data arrived with no pending write available', response);
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:195]: Same here.

File: src/managedwriter/stream_connection.ts [hunk 19]
```diff
@@ -86,7 +87,22 @@ export class StreamConnection extends EventEmitter {
...
     this._connection.on('close', () => {
-      this.trace('connection closed');
+      this.trace('connection closed', this._lastConnectionError);
+      if (this.hasPendingWrites()) {
+        const retrySettings = this._writeClient['_retrySettings'];
+        if (
+          this.isRetryableError(this._lastConnectionError) &&
+          retrySettings.enableWriteRetries
+        ) {
+          this.reconnect();
```
Comment (yirutang) [inline src/managedwriter/stream_connection.ts:97]: Add a log that we are about to reconnect with X number of pending writes.

File: src/managedwriter/stream_connection.ts (hunk 17 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:193]: those `trace` calls are only on DEBUG mode.

File: src/managedwriter/stream_connection.ts [hunk 20]
```diff
@@ -357,7 +357,11 @@ export class StreamConnection extends EventEmitter {
    * Re open appendRows BiDi gRPC connection.
    */
   reconnect() {
-    this.trace('reconnect called');
+    this.trace(
+      'reconnect called with',
```
Comment (shollyman) [inline src/managedwriter/stream_connection.ts:361]: nit: spacing here looks odd, unless node concat adds spaces between fragments.

Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:361]: `console.log` adds spaces, but users can provide other loggers, so I'll use template strings here and add the spaces

File: src/managedwriter/stream_connection.ts (hunk 17 above)
Comment (yirutang) [reply src/managedwriter/stream_connection.ts:193]: Is there a LOG mode? I think those reconnect errors need to be in LOG mode. Even for debug mode, per request logging might be too much, but I am less concerned.

File: src/managedwriter/writer.ts [hunk 21]
```diff
@@ -160,7 +160,7 @@ export class Writer {
     offsetValue?: IInt64Value['value']
   ): PendingWrite {
     let offset: AppendRowRequest['offset'];
-    if (offsetValue) {
+    if (offsetValue || offsetValue === 0) {
```
Comment (yirutang) [inline src/managedwriter/writer.ts:163]: Why is this needed?

File: src/managedwriter/stream_connection.ts [hunk 22]
```diff
@@ -238,13 +253,38 @@ export class StreamConnection extends EventEmitter {
...
   private getNextPendingWrite(): PendingWrite | null {
     if (this._pendingWrites.length > 0) {
-      return this._pendingWrites[0];
+      return this._pendingWrites[this._pendingWrites.length - 1];
     }
     return null;
   }
 
+  private resendAllPendingWrites() {
+    const pendingWrites = [...this._pendingWrites]; // copy array;
```
[the remark this answers is not in the document]
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:268]: this is needed because now when resending the requests, it is added to the `pendingWrites` queue right away instead of waiting for the `write` call to finish. This was causing a infinite loop as the queue was getting items removed (with pop() call) and added right away, so the queue was never empty and the while loop was never ending. 

so we create a copy of the queue, resend requests based on the copy and also empty the internal queue at the same time

File: src/managedwriter/writer.ts (hunk 21 above)
Comment (alvarowolfx) [reply src/managedwriter/writer.ts:163]: because when sending explicitly 0 as offset, this was evaluating as `false` and not being send

File: system-test/managed_writer_client_test.ts [hunk 23]
```diff
@@ -857,6 +865,280 @@ describe('managedwriter.WriterClient', () => {
...
+          });
+
+          let numCalls = 0;
+          let numSucess = 0;
+          const conn = connection['_connection'] as gax.CancellableStream;
+          sandbox
+            .stub(conn, 'write')
+            .callsFake(
+              (
+                chunk: any,
```
Comment (feywind) [inline system-test/managed_writer_client_test.ts:1081]: You can probably do `chunk: unknown` here, and the linter won't complain, and the `as` below should still work.

**Review by feywind — APPROVED** [review]
Nothing super important, I think...

File: src/managedwriter/stream_connection.ts [hunk 24]
```diff
@@ -106,62 +122,53 @@ export class StreamConnection extends EventEmitter {
...
     }
     this.emit('error', err);
   };
 
-  private shouldReconnect(err: gax.GoogleError): boolean {
-    const reconnectionErrorCodes = [
-      gax.Status.UNAVAILABLE,
-      gax.Status.RESOURCE_EXHAUSTED,
+  private handleRetry(err: gax.GoogleError) {
+    const retrySettings = this._writeClient['_retrySettings'];
```
[the remark this answers is not in the document]
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:142]: the problem is that removing the `private` would not protect us from users accessing/modifying it. Although to be honest, they can just do `writeClient['_retrySettings']` same way that I'm doing haha Not sure what to do here

File: src/managedwriter/stream_connection.ts (hunk 17 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:193]: gonna be the first user of that new logger @feywind 🚀

File: src/managedwriter/stream_connection.ts [hunk 25]
```diff
@@ -279,23 +320,26 @@ export class StreamConnection extends EventEmitter {
...
-    if (this._connection.destroyed || this._connection.closed) {
+    if (this.isConnectionClosed()) {
       this.reconnect();
     }
     this.trace('sending pending write', pw);
     try {
-      this._connection.write(request, err => {
+      const request = pw.getRequest();
+      this._pendingWrites.unshift(pw);
+      this._connection?.write(request, err => {
```
[the remark this answers is not in the document]
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:337]: yeah, it's checked by `isConnectionClosed`. Before is was being check explicitly. I'll add a comment.

File: src/managedwriter/stream_connection.ts [hunk 26]
```diff
@@ -86,7 +87,22 @@ export class StreamConnection extends EventEmitter {
...
+      if (this.hasPendingWrites()) {
+        const retrySettings = this._writeClient['_retrySettings'];
+        if (
+          this.isRetryableError(this._lastConnectionError) &&
+          retrySettings.enableWriteRetries
+        ) {
+          this.reconnect();
+          this.resendAllPendingWrites();
+        } else {
+          const err = new Error(
```
Comment (leahecole) [inline src/managedwriter/stream_connection.ts:100]: Consider the [GoogleError class](https://github.com/googleapis/gax-nodejs/blob/80ca2de7cf15c3a547d80b8eccabbcde72929116/gax/src/googleError.ts#L24) from gax possibly?

File: src/managedwriter/stream_connection.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (leahecole) [reply src/managedwriter/stream_connection.ts:142]: 1. Per [this AIP](https://google.aip.dev/client-libraries/4221#retry-configuration-mechanisms) users should be able to configure this - will they be able to configure it if it's private?
2. What is retrySettings comprised of? Is it like the first part of the [RetryOptions class](https://github.com/googleapis/gax-nodejs/blob/80ca2de7cf15c3a547d80b8eccabbcde72929116/gax/src/gax.ts#L80-L82)? (retrycodes and backoff settings)

File: src/managedwriter/stream_connection.ts [hunk 27]
```diff
@@ -106,62 +122,53 @@ export class StreamConnection extends EventEmitter {
...
     this.emit('error', err);
   };
 
-  private shouldReconnect(err: gax.GoogleError): boolean {
-    const reconnectionErrorCodes = [
-      gax.Status.UNAVAILABLE,
-      gax.Status.RESOURCE_EXHAUSTED,
+  private handleRetry(err: gax.GoogleError) {
+    const retrySettings = this._writeClient['_retrySettings'];
+    if (this.isRetryableError(err) && retrySettings.enableWriteRetries) {
```
Comment (leahecole) [inline src/managedwriter/stream_connection.ts:143]: teeny nit - it might be ever so slightly better to swap these so that you check if retries are enabled first then do the check on the error code

File: src/managedwriter/writer_client.ts [hunk 28]
```diff
@@ -69,6 +74,10 @@ export class WriterClient {
       connectionList: [],
     };
     this._open = false;
+    this._retrySettings = {
+      enableWriteRetries: false,
+      maxRetryAttempts: 4,
```
Comment (leahecole) [inline src/managedwriter/writer_client.ts:79]: Why these values? Can they be overridden?

File: src/managedwriter/stream_connection.ts [hunk 29]
```diff
@@ -106,62 +122,53 @@ export class StreamConnection extends EventEmitter {
...
       return;
     }
     this.emit('error', err);
   };
 
-  private shouldReconnect(err: gax.GoogleError): boolean {
-    const reconnectionErrorCodes = [
-      gax.Status.UNAVAILABLE,
-      gax.Status.RESOURCE_EXHAUSTED,
+  private handleRetry(err: gax.GoogleError) {
```
Comment (leahecole) [inline src/managedwriter/stream_connection.ts:141]: two questions
1) does this handle timeout logic at all? 
2) Would you consider modifying gax instead of creating more custom retry logic?

File: src/managedwriter/stream_connection.ts (hunk 26 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts]: changed to `gax.GoogleError` with status code `ABORTED`

File: src/managedwriter/stream_connection.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts]: @leahecole:
1. users can configure it with the methods `enableWriteRetries` and `setMaxRetryAttempts`.
2. I initially though on using `gax.RetryOptions`, but as the retry logic right now is customized to check the number of attempts that a given AppendRequest was retries, it doesn't match what is available right now on `RetryOptions`. But as we improve the retry mechanism to add backoff, I plan to reuse that class.

File: src/managedwriter/stream_connection.ts (hunk 29 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts:142]: @leahecole:
1. when a timeout happen, what is the status code on gax ? I think it is `ABORTED` right ? which is this case is retryable.
2. Yes we considered, but there is custom retry logic here that is tightly coupled with the BQ Storage service, like. triggering a reconnect or not to the service. Other service libraries would definitely be benefited by adding support for retries on BiDi streaming, but we would still have to add some logic here that is specific for BQ Storage.

File: src/managedwriter/stream_connection.ts (hunk 27 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts]: good one.

File: src/managedwriter/writer_client.ts (hunk 28 above)
Comment (alvarowolfx) [reply src/managedwriter/writer_client.ts:84]: they can be overwritten via the `enableWriteRetries` and `setMaxRetryAttempts` methods. The default values came from the Go ManagedWriter - https://github.com/googleapis/google-cloud-go/blob/5a8c88956ea11eaf3e8aa8cffadfca06bf58c51d/bigquery/storage/managedwriter/retry.go#L31

**Review by alvarowolfx — COMMENTED** [review]
answered some question and pushed updates to the PR cc @leahecole @feywind

File: src/managedwriter/stream_connection.ts (hunk 29 above)
Comment (leahecole) [reply src/managedwriter/stream_connection.ts:142]: Timeout is `DEADLINE_EXCEEDED`! ⏰ Re #2, sounds good for now

File: src/managedwriter/stream_connection.ts [hunk 30]
```diff
@@ -86,7 +87,23 @@ export class StreamConnection extends EventEmitter {
...
+          retrySettings.enableWriteRetries &&
+          this.isRetryableError(this._lastConnectionError)
+        ) {
+          this.reconnect();
+          this.resendAllPendingWrites();
+        } else {
+          const err = new gax.GoogleError(
+            'aborted due to failed connection, please retry the request'
+          );
+          err.code = gax.Status.ABORTED;
```
Comment (leahecole) [inline src/managedwriter/stream_connection.ts:103]:
```suggestion
          err.code = gax.Status.UNAVAILABLE;
```

I think this should be unavailable per [this case chart](https://grpc.github.io/grpc/core/md_doc_statuscodes.html). I really had to think about it though - `ABORTED`, `FAILED_PRECONDITION` and `UNAVAILABLE` are so similar, but I think `ABORTED` is closer to an HTTP 409 whereas `UNAVAILABLE` is more like a 503, which is this use case

File: src/managedwriter/stream_connection.ts [hunk 31]
```diff
@@ -86,7 +87,23 @@ export class StreamConnection extends EventEmitter {
...
+        const retrySettings = this._writeClient._retrySettings;
+        if (
+          retrySettings.enableWriteRetries &&
+          this.isRetryableError(this._lastConnectionError)
+        ) {
+          this.reconnect();
+          this.resendAllPendingWrites();
+        } else {
+          const err = new gax.GoogleError(
+            'aborted due to failed connection, please retry the request'
```
Comment (leahecole) [inline src/managedwriter/stream_connection.ts:101]:
```suggestion
            'Connection failure, please retry the request'
```

nit - modifying this language given my suggestion to change the status code

File: src/managedwriter/stream_connection.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (feywind) [reply src/managedwriter/stream_connection.ts]: > the problem is that removing the `private` would not protect us from users accessing/modifying it. Although to be honest, they can just do `writeClient['_retrySettings']` same way that I'm doing haha Not sure what to do here

My general take is that in a JS-based language, it's impossible to keep people from messing with the internals of anything. Especially once you escape TS, it's really easy. So on Pub/Sub I take a policy-based approach: I mark it `@private` and not-for-users as much as I can, and if someone does it anyway, the "you broke it, you buy it" policy applies 😹 (no fixes/bug investigations/API breakage fixes).

I do wish TS had some kind of "public only within the package" marker for easier guardrails, but that's harder.

File: src/managedwriter/stream_connection.ts (hunk 30 above)
Comment (alvarowolfx) [reply src/managedwriter/stream_connection.ts]: applied both changes to message and status code. Thanks for the tips and improvements on that
